### ISSUE TITLE
clangd won't send "documentation" when the definition of a struct is in a header file

### ISSUE THREAD

**Issue by wind0204** [clangd/clangd#164]
Hi, I'm using clangd 9.0.0 as a backend for the code completion and linting in my development with vim and YouCompleteMe. I'd like to ask if this behaviour is normal, clangd returns every useful piece of information for struct members but "documentation" when YouCompleteMe requests a "textDocument/completion" for the completion at the end of the C code line "ptr_to_struct->" and the definition of the struct is in a header file and the header file is included at the top of a C file.

clangd does send the docstring information when the struct is defined in the same file where the ptr_to_struct (struct STRUCT_NAME *) is declared, or when the #include line for the struct definition is in the function scope where the ptr_to_struct is declared.

What I expect is clangd should also send "documentation" along with other information for the docstring I put for the struct members.

Here's an example:

test.h:
```
/**
 * This docstring gets displayed fine :)
 */
struct test {
  int member;  ///< This docstring wouldn't be sent
};
```

main.c:
```
#include "test.h"
#include <stdlib.h>

int main() {
  struct test * test = malloc(sizeof(struct test));
  test->membe  // You won't get "documentation" pieces for a completion on this position
}
```

another_main.c:
```
#include <stdlib.h>

int main() {
#include "test.h"
  struct test * test = malloc(sizeof(struct test));
  test->membe  // You'll get the docstring here in this case.
}
```

Sorry for my bad English!

**Comment by kadircet** [comment]
yes this is a known down-side of documentation in code completion.

Normally we merge documentation coming from our index and AST.
But during code completion, to speed-up the index lookups:
- we only search in a subset of symbols, and members are not in this subset. therefore we can't get the documentation through index.
- we do not deserialize the whole Preamble, therefore documentation cannot be retrieved for members if they are outside the main file.

So you we end up with no documentations at all :/

**Comment by hokein** [comment]
We can get the documentation from the index. I had a patch https://reviews.llvm.org/D56492 doing this, but it is unfinished. I'll pick it up when I have a free slot.

**Comment by wind0204** [comment]
One of the developers of YouCompleteMe said that it's not easy to implement a feature which would send another query for the documentation of currently selected item and update the preview window if the vim session still has the window on, when I asked if it would be easy. I could've tried to implement such a feature in YCM but gave up.

As a person with no knowledge about clangd I'm curious to ask, Does clangd 9.0 have, or would the next version of clangd have a kind of advanced function that would push the detailed information later to hide the latency? With which the user would see the list as soon as possible and still get documentation about struct members.

**Comment by robozati** [comment]
Is there any update on this?

I could work on this issue, but it would be my first time in this codebase so it would take some time.

**Comment by HighCommander4** [comment]
This was fixed in clangd 20 by https://github.com/llvm/llvm-project/pull/120099.
